src/Symfony/Component/HttpKernel/Kernel.php

```diff
@@ -515,6 +517,100 @@ protected function initializeBundles()
+        foreach ($bundles as $bundle) {
+            $bundleFQN = get_class($bundle);
+            $children[$bundleFQN] = BundleDependenciesInterface::DEP_REQUIRED;
+            $rootBundles[$bundleFQN] = $bundle;
+
+            if ($bundle instanceof BundleDependenciesInterface) {
+                $hasDependencies = true;
+            }
+        }
+
+        if (!$hasDependencies) {
+            return $bundles;
+        }
+
+        // Rebuild bundles order with dependencies recursively
+        $bundles = array();
+        $stack = array(get_class($this) => true);
+        $this->appendDependenciesRecursively(
+            $children,
+            $bundles,
+            $rootBundles,
+            $stack
+       );
+
+        return $bundles;
+    }
+
+    /**
+     * Append dependencies of bundles recursively.
+     *
+     * Accepted arguments are as documented below where `string` implies a string with class FQN(Fully Qualified Name),
+     * this string must be in exactly same format as returned by get_class() and PHP 5.5's CLASS constant.
+     * Example of FQN string: "Symfony\Component\HttpKernel\Bundle\BundleInterface"
+     *
+     * @link BundleDependenciesInterface For further details on dependencies and DEP_* constants.
+     *
+     * @param mixed[string]           $children    Dependencies to apply, key FQN, value {@see BundleDependenciesInterface} constants
+     * @param BundleInterface[string] $bundles     Ordered bundles to append dependencies to
+     * @param BundleInterface[string] $rootBundles Loaded Bundles from registerRootBundles()
+     * @param bool[string]            $stack       For recursion protection, and for debug use on exceptions
+     *
+     * @throws \Symfony\Component\HttpKernel\Exception\DependencyMismatchException On missing dependencies
+     */
+    protected function appendDependenciesRecursively(array $children, array &$bundles, array $rootBundles, array $stack)
+    {
+        foreach ($children as $dependencyFQN => $requiredFlag) {
+            if (isset($bundles[$dependencyFQN])) {
+                continue;
+            } elseif (isset($stack[$dependencyFQN])) {
+                throw new DependencyMismatchException(sprintf('Recursive dependency for "%s"', $dependencyFQN), array_keys($stack));
+            }
+
+            // If already loaded root bundle, use that to not re instantiate
+            if (isset($rootBundles[$dependencyFQN])) {
+                $dependency = $rootBundles[$dependencyFQN];
+            } else {
+                if (!class_exists($dependencyFQN)) {
+                    if ($requiredFlag === BundleDependenciesInterface::DEP_OPTIONAL) {
+                        continue;
+                    }
+                    throw new DependencyMismatchException(sprintf('Could not find "%s"', $dependencyFQN), array_keys($stack));
+                }
+                $dependency = new $dependencyFQN();
```

Comment:
Some bundles might have constructor parameters. For example: SonataUserBundle('FOSUserBundle')